src/mem/arena.rs

```diff
@@ -337,14 +361,135 @@ unsafe impl<'arena> Arena for BumpArena<'arena> {
     }
 }
 
+/// A type that can serve as a buffer an arena can allocate from.
+///
+/// # Safety
+///
+/// This function is unsafe to implement. In particular, the following
+/// must hold:
+/// - The value of either method does not change when either function of this
+///   trait is called, unless `self` is moved.
+/// - `slice::from_raw_parts_mut(as_ptr(), len())` is a valid reference to
+///   a slice.
+/// - Calling either function does not create a reference to is pointee, to
+///   avoid violating the aliasing rules.
+pub unsafe trait Buf {
+    /// Returns a pointer to the buffer of bytes managed by this `Buf`.
+    ///
+    /// This function may be called while there exist outstanding references
+    /// into this buffer.
+    #[allow(clippy::wrong_self_convention)]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8;
+
+    /// Returns the length of the buffer, in bytes.
+    fn len(this: &UnsafeCell<Self>) -> usize;
+}
+
+unsafe impl Buf for [u8] {
+    #[inline]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8 {
+        // SAFETY: Casting a fat pointer to a thin pointer always pulls out
+        // the "pointer" part.
+        this.get() as *mut u8
+    }
+
+    #[inline]
+    fn len(this: &UnsafeCell<Self>) -> usize {
+        unsafe {
+            // TODO: This currently materializes a reference to the slice.
+            // We would like to use `<*mut [u8]>::len()`, but that requires
+            // nightly. =(
+            (&*this.get()).len()
+        }
+    }
+}
+
+unsafe impl<const N: usize> Buf for [u8; N] {
+    #[inline]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8 {
+        // SAFETY: *mut [T; N] as *mut T is a pointer to the first element.
+        this.get() as *mut u8
+    }
+
+    #[inline]
+    fn len(_: &UnsafeCell<Self>) -> usize {
+        N
+    }
+}
+
+// SAFETY: We can materialize the &mut Vec without any trouble, since the
+// aliased memory is the data it points to, which is *not* materialized
+// unless Deref is called, which neither of the functions below do.
+#[cfg(feature = "std")]
+unsafe impl Buf for Vec<u8> {
+    #[inline]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8 {
+        unsafe { (&mut *this.get()).as_mut_ptr() }
+    }
+
+    #[inline]
+    fn len(this: &UnsafeCell<Self>) -> usize {
+        unsafe { (&*this.get()).len() }
+    }
+}
+
+// SAFETY: Use the fact that Box[T] is representation-equivalent to
+// &mut T (due to FFI safety).
+#[cfg(feature = "std")]
+unsafe impl Buf for Box<[u8]> {
+    #[inline]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8 {
+        unsafe {
+            Buf::as_ptr(&*(this as *const _ as *const UnsafeCell<&mut [u8]>))
```

Comment:
this feels iffy to me; is it safe to materialize an `&UnsafeCell<&mut [u8]>` here?

Reply:
Made irrelevant. =)